proxy/proxy.go

```diff
@@ -401,8 +401,20 @@ func (p *Proxy) Resolve(d *DNSContext) error {
 	// Use cache only if it's enabled and the query doesn't use custom
 	// upstreams.
 	cacheWorks := p.cache != nil && d.CustomUpstreamConfig == nil
-	if cacheWorks {
-		if p.replyFromCache(d) {
+	if cacheWorks && !d.disableCacheReading {
+		hit, isNotExpired := p.replyFromCache(d)
+		hit = hit && (p.Config.CachePrefetch || isNotExpired)
+		if hit {
+			if p.Config.CachePrefetch && !isNotExpired {
```

Comment:
I feel like the code is getting uncomfortably deeply-nested here.  Perhaps factor out into a separate method?